libsplinter/src/network/ref_map.rs

```diff
@@ -0,0 +1,215 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::collections::HashMap;
+use std::{error, fmt};
+
+/// A map of that will keep track of the number of times an id has been added, and only remove the
+/// id once the reference count is 0.
+///
+/// The RefMap also keeps track of redirects. This allows ids to be changed while keeping the same
+/// refefrence count.
+pub struct RefMap {
+    // id to reference count
+    references: HashMap<String, u64>,
+    redirects: HashMap<String, String>,
+}
+
+impl RefMap {
+    pub fn new() -> Self {
+        RefMap {
+            references: HashMap::new(),
+            redirects: HashMap::new(),
+        }
+    }
+
+    pub fn add_ref(&mut self, id: String) -> u64 {
+        // check if id is
+        let ref_id = {
+            if !self.references.contains_key(&id) {
+                // if the the id is for an old reference, find updated id
+                if let Some(ref_id) = self.redirects.get(&id) {
+                    ref_id.clone()
+                } else {
+                    // if the id is not in the reference or redirects, it is a new reference
+                    id
+                }
+            } else {
+                id
+            }
+        };
+
+        if let Some(ref_count) = self.references.remove(&ref_id) {
+            let new_ref_count = ref_count + 1;
+            self.references.insert(ref_id, new_ref_count);
+            new_ref_count
+        } else {
+            self.references.insert(ref_id, 1);
+            1
+        }
+    }
+
+    pub fn update_ref(&mut self, old_id: String, new_id: String) -> Result<(), RefUpdateError> {
+        if let Some(ref_count) = self.references.remove(&old_id) {
+            self.references.insert(new_id.clone(), ref_count);
+
+            // update the old forwards
+            for (_, v) in self.redirects.iter_mut().filter(|(_, v)| **v == old_id) {
+                *v = new_id.clone()
+            }
+
+            self.redirects.insert(old_id, new_id);
+
+            Ok(())
+        } else {
+            Err(RefUpdateError { id: new_id })
+        }
+    }
+
+    // remove_ref, return id if the peer id was removed
+    pub fn remove_ref(&mut self, id: &str) -> Option<String> {
+        // check if id is
+        let ref_id = {
+            if !self.references.contains_key(id) {
+                // if the the id is for an old reference, find updated id
+                if let Some(ref_id) = self.redirects.get(id) {
+                    ref_id.to_string()
+                } else {
+                    // if the id is not in the reference or redirects, the reference does not exist
+                    panic!("Trying to remove a reference that does not exist: {}", id)
```

Comment:
@vaporos had mentioned that we should panic if we try to remove a ref that does not exist. Where that panic should be raised could be changed but this refmap will also be used the ServiceManager as well so I though it would be good to happen at this layer.

Reply:
If this is only going to be used by the peer and service managers then that's probably fine. May want to specifically call out that this panics in a docstring though.

Reply:
added a comment